File: calculator/calculator.cpp
#include <cctype>
#include <iostream>
#include <stdexcept>
#include <string>

using namespace std;

string INT = "INT", PLUS = "PLUS", MINUS = "MINUS", MUL = "MUL", DIV = "DIV", LP = "LP", RP = "RP",
       EOS = "EOS";

class Token {
public:
  Token(const string &kind, const string &value)
      : m_kind(kind), m_value(value) {}

  string getKind() { return m_kind; }

  string getValue() { return m_value; }

private:
  string m_kind;
  string m_value;
};

class Lexer {
public:
  Lexer(string &text)
      : m_text(text), pos(0), currentChar(text[pos]), maxLen(text.size()) {}

  Token getNextToken() {
    if (pos >= maxLen) {
      return Token(EOS, "\0"); // todo: check this
    }

    if (isdigit(currentChar)) {
      string intToken = getInt();
      return Token(INT, intToken);
    }

    if (isspace(currentChar)) {
      skipSpace();
      return getNextToken();
    }

    if (currentChar == '+') {
      advance();
      return Token(PLUS, "+");
    }

    if (currentChar == '-') {
      advance();
      return Token(MINUS, "-");
    }

    if (currentChar == '*') {
      advance();
      return Token(MUL, "*");
    }

    if (currentChar == '/') {
      advance();
      return Token(DIV, "/");
    }

    if (currentChar == '(') {
      advance();
      return Token(LP, "(");
    }

    if (currentChar == ')') {
      advance();
      return Token(RP, ")");
    }

    throw runtime_error("Invalid character!");
  }

private:
  bool advance() {
    pos++;
    if (pos < maxLen) {
      currentChar = m_text[pos];
      return true;
    } else {
      return false;
    }
  }

  void skipSpace() {
    while (isspace(currentChar)) {
      advance();
    }
  }

  string getInt() {
    string res = "";
    while (isdigit(currentChar)) {
      res += currentChar;
      if (!advance()) {
        break;
      }
    }
    return res;
  }

  const string m_text;
  int pos;
  const size_t maxLen;
  char currentChar;
};

class Interpreter {
public:
  Interpreter(Lexer & lexer) : m_lexer(lexer), currentToken(m_lexer.getNextToken()) {}

  int divisor(int fac) {
    if (fac == 0) {
      throw runtime_error("Dividing by 0!");
    } else {
      return fac;
    }
    
  }

  void eat(string kind) {
    if (currentToken.getKind() == kind) {
      currentToken = m_lexer.getNextToken();
    } else {
      error();
    }
  }

  int factor() {
    if (currentToken.getKind() == INT) {
      string val = currentToken.getValue();
      eat(INT);
      return stoi(val);
    } else if (currentToken.getKind() == LP) {
      eat(LP);
      int result = expr();
      eat(RP);
      return result;
    }
    error();
  }

  int term() {
    int result = factor();

    while (currentToken.getKind() == MUL || currentToken.getKind() == DIV) {
      string op = currentToken.getKind();

      if (op == MUL) {
        eat(MUL);
        result *= factor();
      }

      if (op == DIV) {
        eat(DIV);
        int fac = factor();
        if (fac == 0) {
            throw runtime_error("Dividing by 0!");
          }
        result /= factor();
      }
    }
    return result;
  }

  int expr() {
    int result = term();

    while (currentToken.getKind() == PLUS || currentToken.getKind() == MINUS) {
      string op = currentToken.getKind();

      if (op == PLUS) {
        eat(PLUS);
        result += term();
      }

      else if (op == MINUS) {
        eat(MINUS);
        result -= term();
      }
    }
    return result;
  }

  void error() { throw runtime_error("Invalid syntax."); }

private:
  Lexer m_lexer;
  Token currentToken;
};

int main() {
  string prompt = "calc>";
  string text;
  while (true) {
    cout << prompt << endl;
    getline(cin, text);
    Lexer lexer(text);
    Interpreter interpreter(lexer);
    cout << interpreter.expr() << endl;
  }
  return 0;
}

File: calculator/calculator.py

INTEGER, PLUS, MINUS, TIMES, DIV, LPAR, RPAR, EOS = "INTEGER", "PLUS", "MINUS", "TIMES", "DIV", "LPAR", "RPAR", "EOS"


class Token:
    """
    Stores a token as a (type, value) pair.
    """
    def __init__(self, type, value):
        self.type = type
        self.value = value

    def __str__(self):
        return f"Token({self.type}, {self.value})"

    def __repr__(self):
        return self.__str__()


class Lexer:
	"""
	Tokenizes input.
	Call `get_next_token` to get next token in input
	"""
	def __init__(self, text):
		self.text = text
		self.pos = 0
		self.max_len = len(text)
		self.current_char = self.text[self.pos]

	def error(self):
		raise Exception("Invalid character.")

	def advance(self):
		self.pos += 1
		if self.pos < self.max_len:
			self.current_char = self.text[self.pos]
		else:
			self.current_char = None

	def skip_space(self):
		while self.current_char.isspace():
			self.advance()

	def get_int(self):
		int_str = ""
		while self.current_char is not None and self.current_char.isdigit():
			int_str += self.current_char
			self.advance()
		return int(int_str)

	def get_next_token(self):
		if self.pos >= self.max_len:
			return Token(EOS, None)

		if self.current_char.isspace():
			self.skip_space()
			return self.get_next_token()

		if self.current_char.isdigit():
			return Token(INTEGER, self.get_int())

		if self.current_char == "+":
			self.advance()
			return Token(PLUS, "+")

		if self.current_char == "-":
			self.advance()
			return Token(MINUS, "-")

		if self.current_char == "*":
			self.advance()
			return Token(TIMES, "*")

		if self.current_char == "/":
			self.advance()
			return Token(DIV, "/")
		
		if self.current_char == "(":
			self.advance()
			return Token(LPAR, "(")
		
		if self.current_char == ")":
			self.advance()
			return Token(RPAR, ")")

		self.error()

class Interpreter:
	"""
	Contains a parser and interpreter.
	Parser structures stream of tokens produced by lexer.
	Interpreter computes the structure given by the parser
	"""
	def __init__(self, lexer):
		self.lexer = lexer
		self.pos = 0
		self.current_token = self.lexer.get_next_token()

	def error(self):
		raise Exception("Invalid syntax.")

	def eat(self, type):
		if self.current_token.type == type:
			self.current_token = self.lexer.get_next_token()
		else:
			raise self.error()

	def factor(self):
		"""
		factor : INTEGER | LPAR expr RPAR
		"""
		if self.current_token.type == INTEGER:
			result = self.current_token.value
			self.eat(INTEGER)
		elif self.current_token.type == LPAR:
			self.eat(LPAR)
			result = self.expr()
			self.eat(RPAR)
		
		return result

	def term(self):
		"""
		term : factor((TIMES | DIV) factor)*
		"""
		result = self.factor()

		while self.current_token.type in (TIMES, DIV):
			token = self.current_token
			if token.type == TIMES:
				self.eat(TIMES)
				result *= self.factor()
			if token.type == DIV:
				self.eat(DIV)
				div = self.factor()
				if div == 0:
					raise ZeroDivisionError("Dividing by 0!")
				result /= self.factor()
		return result

	def expr(self):
		"""
		expr : term((PLUS | MINUS) term)*
		"""
		result = self.term()

		while self.current_token.type in (PLUS, MINUS):

			token = self.current_token
			if token.type == PLUS:
				self.eat(PLUS)
				result += self.term()
			if token.type == MINUS:
				self.eat(MINUS)
				result -= self.term()

		return result



def main():
    while True:
        try:
            text = input("calc>")
        except EOFError:
            print("Invalid input")
            break
        if not text:
            continue
        lexer = Lexer(text)
        interpreter = Interpreter(lexer)
        result = interpreter.expr()
        print(result)


if __name__ == "__main__":
    main()
